### PR TITLE
Add caching for broker responses in memory, lookup by origin and thumbprint

### PR DESCRIPTION
**Pull request by pkanher617**
This PR adds the ability for the broker to cache responses from a redirect. It also uses embedded app origin to look up these responses, preventing child apps from receiving responses that do not belong to them.

### CODE REVIEW

File: lib/msal-browser/src/app/ClientApplication.ts
Comment (tnorling) [inline lib/msal-browser/src/app/ClientApplication.ts:319]:
```suggestion
            const redirectStartPage = userRedirectStartPage || window.location.href;
```

File: lib/msal-browser/src/broker/client/BrokerClientApplication.ts
Comment (tnorling) [inline lib/msal-browser/src/broker/client/BrokerClientApplication.ts:47]: This feels a little hacky but I can't think of an alternative at the moment. This may be something we should do in the base `handleRedirectPromise` to prevent the race conditions we hear about from time to time.

File: lib/msal-browser/src/config/Configuration.ts
Comment (tnorling) [inline lib/msal-browser/src/config/Configuration.ts:60]: Any reason a brokered scenario would not want to use the same `redirectStartPage` and `onRedirectNavigate` as the non-brokered scenario?

Comment (pkanher617) [reply lib/msal-browser/src/config/Configuration.ts:60]: `redirectStartPage` and `onRedirectNavigate` shouldn't be set by the embedded application, since they are not the ones doing the redirect. The broker app should specify A. which page the redirect starts on and B. what happens after a redirect.

Comment (pkanher617) [reply lib/msal-browser/src/config/Configuration.ts:60]: We block all redirect APIs in an embedded scenario anyway, so these config options don't get used today in an embedded scenario.

File: lib/msal-browser/src/broker/client/BrokerClientApplication.ts
Comment (pkanher617) [reply lib/msal-browser/src/broker/client/BrokerClientApplication.ts:47]: It is a bit hacky, I needed a way to wait for the response to finish in case the broker receives a request for the redirect response before it is done trading for the tokens. Happy to hear alternatives. The reason I haven't done this in the base `handleRedirectPromise()` is because I am trying to keep broker specific code away from the base functions.

Comment (tnorling) [reply lib/msal-browser/src/broker/client/BrokerClientApplication.ts:47]: I'm advocating for making this not broker specific. We have scenarios where if you call `handleRedirectPromise` twice at the same time it has the potential to create a race condition and break. I think moving this logic up one level, into the base function would solve both problems. Ultimately what you're trying to prevent here is having more than one instance of `handleRedirectPromise` running concurrently, right?

File: lib/msal-browser/src/config/Configuration.ts
Comment (tnorling) [reply lib/msal-browser/src/config/Configuration.ts:60]: Sorry should have been more clear, why do we need separate `redirectStartPage` and `brokerRedirectStartPage` params instead of having the broker just use `redirectStartPage` regardless of whether it is getting tokens for itself or for an embedded app?

File: lib/msal-browser/src/broker/client/BrokerClientApplication.ts
Comment (pkanher617) [reply lib/msal-browser/src/broker/client/BrokerClientApplication.ts:47]: That is more of a side effect, the main reason is that I need a way to determine if we are currently trading for tokens or not, so that when an embedded app makes a request we can wait for the tokens to be retrieved before returning a response back from the broker.

File: lib/msal-browser/src/config/Configuration.ts
Comment (pkanher617) [reply lib/msal-browser/src/config/Configuration.ts:60]: redirectStartPage is set in the context of the client making the request, and is only a parameter on the request. The broker doesn't have any control over what the embedded application sends as the redirectStartPage, but we don't want the embeddedApp telling the broker to use the redirectStartPage for its own app - the broker should specify that themselves. The easiest way to do that is to add a broker config option for it in the constructor config.

Comment (tnorling) [reply lib/msal-browser/src/config/Configuration.ts:60]: Got it. Forgot the params are only set on the request. Makes sense, thanks for clarifying

File: lib/msal-browser/src/app/ClientApplication.ts
Comment (jasonnutter) [inline lib/msal-browser/src/app/ClientApplication.ts:359]: Popups opened by the broker should also include this logic.

File: lib/msal-browser/src/broker/client/BrokerClientApplication.ts
Comment (jasonnutter) [inline lib/msal-browser/src/broker/client/BrokerClientApplication.ts:53]:
```suggestion
        return (redirectResponse && !redirectResponse.tokensToCache) ? redirectResponse : null;
```

File: lib/msal-browser/src/broker/client/BrokerClientApplication.ts
Comment (jasonnutter) [inline lib/msal-browser/src/broker/client/BrokerClientApplication.ts:137]: `acquireToken`?

File: lib/msal-browser/src/broker/client/BrokerClientApplication.ts
Comment (jasonnutter) [inline lib/msal-browser/src/broker/client/BrokerClientApplication.ts:178]: Can you factor these into helper functions? Would help readability.

File: lib/msal-browser/src/app/ClientApplication.ts
Comment (pkanher617) [reply lib/msal-browser/src/app/ClientApplication.ts:359]: We will need to add that property to the broker options then, since the broker will need to specify the popup in constructor.

File: lib/msal-browser/src/config/Configuration.ts
Comment (pkanher617) [inline lib/msal-browser/src/config/Configuration.ts:60]: Add Omit type for redirect/popup params

Comment (pkanher617) [reply lib/msal-browser/src/config/Configuration.ts:60]: Think about adding in a later version?

File: lib/msal-browser/src/broker/client/BrokerClientApplication.ts
Comment (pkanher617) [inline lib/msal-browser/src/broker/client/BrokerClientApplication.ts:45]: Add comments

File: lib/msal-browser/src/broker/client/BrokerClientApplication.ts
Comment (pkanher617) [inline lib/msal-browser/src/broker/client/BrokerClientApplication.ts:129]: Add comments

File: lib/msal-browser/src/app/ClientApplication.ts
Comment (pkanher617) [reply lib/msal-browser/src/app/ClientApplication.ts:359]: Will resolve in a separate PR once the fixes for popup flow has been merged to dev

File: lib/msal-browser/src/broker/client/BrokerClientApplication.ts
Comment (pkanher617) [reply lib/msal-browser/src/broker/client/BrokerClientApplication.ts:48]: done.

File: lib/msal-browser/src/broker/client/BrokerClientApplication.ts
Comment (pkanher617) [reply lib/msal-browser/src/broker/client/BrokerClientApplication.ts:132]: done.

File: lib/msal-browser/src/config/Configuration.ts
Comment (pkanher617) [reply lib/msal-browser/src/config/Configuration.ts]: done.

File: lib/msal-browser/src/broker/client/BrokerClientApplication.ts
Comment (pkanher617) [reply lib/msal-browser/src/broker/client/BrokerClientApplication.ts]: done.

File: lib/msal-browser/src/broker/client/BrokerClientApplication.ts
Comment (pkanher617) [reply lib/msal-browser/src/broker/client/BrokerClientApplication.ts]: fixed

File: lib/msal-browser/src/broker/client/BrokerClientApplication.ts
Comment (pkanher617) [reply lib/msal-browser/src/broker/client/BrokerClientApplication.ts]: fixed

File: lib/msal-browser/src/app/ClientApplication.ts
Comment (pkanher617) [reply lib/msal-browser/src/app/ClientApplication.ts]: done.

File: lib/msal-browser/src/app/PublicClientApplication.ts
Comment (tnorling) [inline lib/msal-browser/src/app/PublicClientApplication.ts:6]: Once you update from dev these request types should be imported from within browser, FYI

**Review by tnorling — APPROVED** [review]

File: lib/msal-browser/src/app/PublicClientApplication.ts
Comment (pkanher617) [reply lib/msal-browser/src/app/PublicClientApplication.ts]: done.

**Pull request merged**
